packages/charts/src/components/legend/series/series-horizontal.component.tsx

```diff
@@ -1,56 +1,188 @@
+    0,
+    0,
+  ]);
+
+  const [renderMode, setRenderMode] = useState<RenderMode>('list');
+  const [sliderDimension, setSliderDimension] = useState<[number, number]>([
+    0,
+    0,
+  ]);
+  const [direction, setDirection] = useState(0);
+
+  const [startOffset, endOffset] = dataSeriesOffset;
+
+  useEffect(() => {
+    if (renderMode === 'slider') {
+      const colorsInPallete = colorPalette.length;
+      const containerWidth = containerRef.current.offsetWidth;
+
+      const dimensionCapacity = Math.floor(
+        (containerWidth - 2 * (BUTTON_DIMENSION + BUTTON_SHADOW_SIZE)) /
+          (itemWidth + itemGap)
+      );
+
+      const itemsInSlider =
+        dimensionCapacity > colorsInPallete
+          ? colorsInPallete
+          : dimensionCapacity;
+
+      const sliderWidth =
+        itemsInSlider * (itemWidth + itemGap) +
+        2 * (BUTTON_DIMENSION + BUTTON_SHADOW_SIZE) -
+        itemGap;
+
+      setSliderDimension(([, height]) => [sliderWidth, height]);
+      setDataSeriesOffset([0, itemsInSlider]);
+      setCalculationReady(true);
+    }
+  }, [renderMode]);
+
+  useEffect(() => {
+    const notEnoughColors = dataSeries.length > colorPalette.length;
+    const contentOverflow = hasContentOverflow(
+      'horizontal',
+      containerRef.current
+    );
+
+    if (contentOverflow || notEnoughColors) {
+      const sliderHeight = containerRef.current.offsetHeight;
 
-  const renderSlider = mode === 'slider';
+      setSliderDimension([0, sliderHeight]);
+      setRenderMode('slider');
+    } else {
+      setCalculationReady(true);
+    }
+  }, []);
 
   return (
-    <Card borderPosition="left" {...card}>
-      <div ref={element} style={{ overflow: 'hidden' }}>
-        {renderSlider ? (
-          <Slider mode="horizontal" slidesPerRow={2}>
-            {children}
-          </Slider>
-        ) : (
-          <Layout>
-            {mode === 'list' && children}
-            {mode === 'group' && (
-              <Group groupNode={GroupedSeries}>{children}</Group>
+    <div
+      style={{ background: 'transparent', width: '100%', position: 'relative' }}
+    >
+      <Container ref={containerRef} calculationReady={calculationReady}>
+        <AlignmentContainer alignment={alignment}>
+          <Card borderPosition="left" {...card}>
+            {renderMode === 'list' ? (
+              <Layout itemSpace={itemGap}>
+                {renderNodes(
+                  dataSeries.slice(0, colorPalette.length),
+                  elementRef,
+                  itemWidth
+                )}
+              </Layout>
+            ) : (
+              <Slider
+                mode="horizontal"
+                dimension={sliderDimension}
+                nextDisabled={endOffset === dataSeries.length}
+                previousDisabled={startOffset === 0}
+                direction={direction}
+                animation={(itemIndex) => {
+                  const itemPosition =
```

Comment:
As we discussed - the animation logic become too complex so please, extract it to separate function responsible for creation of transition.

Reply:
Good call 👍  done